### PR TITLE
Added collect all function to find all emojis in a string without eliminating duplicates

### PR DESCRIPTION
**Pull request by mastern2k3**
I really wanted to use gomoji but it's missing a way to collect all emojis in a string as-is, without ignoring repeating emojis.

I am building a service that analyzes emojis in strings and need a way to detect which emojis appear and how many times.

I also feel like this behavior is more suiting the name `FindAll` and the current behavior of `FindAll` should be relocated to a `FindAllDistinct`.

### CODE REVIEW

**Comment by forPelevin** [comment]
Hi mastern2k3!

Thank you for your contribution! 

Yeah, I'll rename the `FindAll` method with the following significant changes. However, it seems overkill to break backward compatibility to rename a method name right now.

File: gomoji.go [hunk 1]
```diff
@@ -82,6 +82,27 @@ func GetInfo(emoji string) (Emoji, error) {
 	return em, nil
 }
 
+// CollectAll finds all emojis in given string. Unlike FindAll, this does not
+// distinct repeating occurrences of emoji.
+func CollectAll(s string) []Emoji {
```
Comment (forPelevin) [inline gomoji.go:87]: Great job! Could you add unit tests, please?

**Comment by mastern2k3** [comment]
> it seems overkill to break backward compatibility to rename a method name right now.
Strongly agree, will add tests soon.

**Comment by mastern2k3** [comment]
When I added the tests I noticed I'm getting duplicate results because some emojis were getting registered from both

```golang
	gr := uniseg.NewGraphemes(s)
	for gr.Next() {
		if em, ok := emojiMap[gr.Str()]; ok {
			emojis[gr.Str()] = em
		}
	}
```

and

```golang
	for _, r := range s {
		if em, ok := emojiMap[string(r)]; ok {
			emojis[string(r)] = em
		}
	}
```

I didn't quite get the difference, it looks like `uniseg.NewGraphemes(s)` picks up all emojis but some just don't appear correctly from `gr.Str()` and don't hit the correct map keys.

File: gomoji.go [hunk 2]
```diff
@@ -82,6 +82,46 @@ func GetInfo(emoji string) (Emoji, error) {
...
+		} else {
+			for _, r := range s[start:end] {
+				if em, ok := emojiMap[string(r)]; ok {
+					emojis = append(emojis, em)
+				}
+			}
+		}
+
+		lastSeen = end
+	}
```
Comment (forPelevin) [inline gomoji.go:114]: The cycle looks hard to get, though. 
Maybe we can try something like this? Will it work for you? It looks like it passes your test cases.

```
func CollectAll(s string) []Emoji {
	var emojis []Emoji

	gr := uniseg.NewGraphemes(s)
	for gr.Next() {
		if em, ok := emojiMap[gr.Str()]; ok {
			emojis = append(emojis, em)
			continue
		}

		start, end := gr.Positions()
		for _, r := range s[start:end] {
			if em, ok := emojiMap[string(r)]; ok {
				emojis = append(emojis, em)
			}
		}
	}

	return emojis
}
```

**Comment by forPelevin** [comment]
> I didn't quite get the difference, it looks like uniseg.NewGraphemes(s) picks up all emojis but some just don't appear correctly from gr.Str() and don't hit the correct map keys.

Yep, you're right. Some emojis are invisible for `uniseg.NewGraphemes(s)` and some for `for range` by runes. So that's why we use both approaches here.

**Comment by forPelevin** [comment]
hey @mastern2k3,

It's been a while since our last interaction. Is everything good? Does my suggestion work for you?

**Comment by mastern2k3** [comment]
Hey @forPelevin all is well, looks like your suggestion should work.
I'm investigating something different, namely - why doesn't `emojiMap[gr.Str()]` always hit?
Looks like this happens only for a specific set of emojis.

They get picked up by the `uniseg.NewGraphemes(s)` iterator, they also appear as emojis when you print them as `gr.Str()`, but when put through `emojiMap` it doesn't hit.

**Comment by forPelevin** [comment]
Hey @mastern2k3.

> I'm investigating something different, namely - why doesn't emojiMap[gr.Str()] hit for all emojis?
> Looks like this happens only for a specific set of emojis.

If you print them by bytes, you'll see hidden bytes. That's why they cannot be found in `emojiMap`. Unfortunately, there is no obvious way to pre-generate all the emojis, so we need to double-check everything in runtime.

**Comment by forPelevin** [comment]
Hi @mastern2k3 

Could you make changes from previos message and I will merge the pull request, please?

**Comment by mastern2k3** [comment]
@forPelevin I fixed it :+1:

**Review by forPelevin — APPROVED** [review]

**Comment by forPelevin** [comment]
Nice job! Thank you for the PR. I've merged it.

**Pull request merged**